Allow covariance for read-only attributes

TH3CHARLie opened the pull request:
resolves #5836

For now, I've not added any tests yet. I've tested against code samples in #5836 and #3260 and the patch is working but adding either of them seems to require modifying a lot in test fixtures

The new helper function `is_writable_attribute` is actually borrowed from `check_no_writable`. I'd also like to refactor the latter function's name and implement it using the new helper in this PR, since I think its name is kind of misleading considering the comments
https://github.com/python/mypy/blob/ea3c65cc377fcb6e810dec9ed6314b25113dcda2/mypy/checker.py#L2391-L2402

JukkaL commented:
> but adding either of them seems to require modifying a lot in test fixtures

What's missing in text fixtures? Maybe you can simplify the examples to not rely on so many features.

TH3CHARLie commented:
> > but adding either of them seems to require modifying a lot in test fixtures
> 
> What's missing in text fixtures? Maybe you can simplify the examples to not rely on so many features.

Good point. I've simplified the case from #5836 and added it. I've also done some refactor work to reuse the new helper added and modify the function's name to make it less misleading. But it is not the main focus of the PR and I'd like to know your thoughts on it.
